Add Rating setting to the Video Block when VideoPress is enabled

jgcaruso opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

Fixes 740-gh-Automattic/greenhouse

~~Depends on https://github.com/Automattic/jetpack/pull/17681 for new endpoint to update the rating.~~ merged!

#### Changes proposed in this Pull Request:
* Adds a new `jetpack_videopress` object field to the v2 api for returning videopress data instead of the one-off `jetpack_videopress_guid`
* This PR adds a new sidebar panel for "Video File Settings" (which affects the video file directly, instead of just the current block)
* A setting for 'Rating' is in the new sidebar panel, that auto-saves when the value is changed (same behaviour as Media Library)

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
phpunit (Tests original field and new field)
`yarn docker:phpunit --filter Test_WPCOM_REST_API_V2_Attachment_VideoPress_`

Manual
1. Apply this PR to your dev environment
2. Run `yarn build-extensions`
3. Ensure you have a paid Jetpack plan that includes video hosting
4. Enable `Enable high-speed, ad-free video player` in Jetpack > Settings > Performance
5. In a post, add a `/video` block and upload a new video
6. Once the video has uploaded, change the value for Rating in the sidebar. It will save automatically.
7. Open the Media Library, find your video and open it
8. The rating in the media library should match the rating you selected on the block
9. Update the rating here to a new value. It will save automatically.
10. Go back to your post and select the video block (if you left the previous window/tab open you may need to refresh the page)
11. The rating should match the rating you selected from the media library
12. View the public page for the post. If you selected a Rating other than `G`, then you should be shown the "Age Gate" before being allowed to play the video.

Editing existing video blocks
1. Select an existing video block and Click the Pencil icon in the toolbar
2. Upload or choose a new video that has a different rating than the current video
3. Observe the Rating in the sidebar, ensure that rating changes to the rating for the new video and that the video preview changes to the new video.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
*

roundhill commented on _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-videopress.php:
```diff
@@ -100,6 +124,30 @@ public function videopress_block_update_meta( $request ) {
 		$response_body = json_decode( wp_remote_retrieve_body( $result ) );
 		if ( is_bool( $response_body ) && $response_body ) {
+
+			if ( ! defined( 'IS_WPCOM' ) || ! IS_WPCOM ) {
```
A comment might be nice here to explain why we had to create separate code for Jetpack or WPCOM

roundhill reviewed (approved):
Tested well on Jetpack and WP.com for me. Added a thought about maybe adding a comment for future devs that run across this code.

roundhill reviewed (approved):
Tested on wp.com and on a test Jetpack site well for me. Rating changes are saved and applied properly.

jgcaruso commented on _inc/lib/core-api/wpcom-fields/attachment-fields-videopress-data.php:
```diff
@@ -0,0 +1,175 @@
+<?php
```
Will do. I was following the other classes named in that folder, but I suppose the rule may be newer than them.

jeherve replied:
Yeah, that's most likely the case.

jgcaruso commented on _inc/lib/core-api/wpcom-fields/attachment-fields-videopress-data.php:
```diff
@@ -0,0 +1,175 @@
+	 * @return false|int
+	 */
+	public function is_video( $attachment ) {
+		return wp_startswith( $attachment->post_mime_type, 'video/' );
```
I added the check, but I'm curious where you were seeing a notice?

jeherve replied:
I did not get any notices in my tests, but I do know that in some cases (large media that the site fails to handle properly on upload), some metadata can be missing from attachments.

jgcaruso commented:
@jeherve I've addressed your comments, but as we're seeing in https://github.com/Automattic/jetpack/pull/18006 there appears to be an issue in `master` that is causing problems with the video block in WP 5.6 . So old video blocks aren't behaving themselves if testing in that environment.

Not sure what that means for this PR?

jeherve commented on _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-videopress.php:
```diff
@@ -100,6 +124,36 @@ public function videopress_block_update_meta( $request ) {
+				}
+
+				if ( isset( $json_params['display_embed'] ) && isset( $meta['videopress']['display_embed'] ) ) {
+					$meta['videopress']['display_embed'] = $json_params['display_embed'];
+					$should_update_meta                  = true;
+				}
+
+				if ( isset( $json_params['rating'] ) && isset( $meta['videopress']['rating'] ) ) {
+					$meta['videopress']['rating'] = $json_params['rating'];
```
I think we should add some validation here, so one doesn't just pass random parameters that won't really mean much.

jgcaruso replied:
Good call. The endpoint it calls prior to saving does validate the value, but if its ever changed to ignore the value instead of returning a failed response this could still save when it shouldn't.

I've refactored some existing rating text validation into a helper function so we're not duplicating the rating strings everywhere.

jeherve reviewed (changes requested):
This works for me. I only have one remark, below.

I don't think we need to worry about block invalidation in this specific PR, since the problem isn't new and will be addressed in another PR.

jeherve commented:
r219218-wpcom
